Sources/SourceKitBazelBSP/RequestHandlers/BuildTargets/BazelTargetStore.swift

```diff
@@ -183,31 +201,47 @@ final class BazelTargetStoreImpl: BazelTargetStore {
             return cachedTargets
         }
 
-        // Start by determining which platforms each top-level app is for.
-        // This will allow us to later determine which sets of flags to provide
-        // depending on the target / platform combo the LSP is interested in,
-        // as well as throwing an error if the user provided something that we
-        // don't currently know how to process.
-        let topLevelTargetData = try bazelTargetQuerier.queryTopLevelRuleTypes(
-            forConfig: initializedConfig,
-            rootUri: initializedConfig.rootUri,
-        )
-        let topLevelTargets = topLevelTargetData.map { $0.0 }
+        let topLevelRuleKinds = Set(TopLevelRuleType.allCases.map(\.rawValue))
+
+        // Query all the targets we are interested in one invocation:
+        //  - Top-level targets (e.g. `ios_application`, `ios_unit_test`, etc.)
+        //  - Dependencies of the top-level targets (e.g. `swift_library`, `objc_library`, etc.)
+        let allTargets =
+            try bazelTargetQuerier
+            .queryTargets(
+                config: initializedConfig,
+                topLevelRuleKinds: topLevelRuleKinds,
+                dependencyKinds: Self.supportedKinds
+            )
+
+        // Find the top-level targets (based on our supported rule kinds) from the query results.
+        var topLevelTargetLabels: [String] = []
+        var topLevelTargetTypes: [TopLevelRuleType] = []
+        for target in allTargets {
+            let kind = target.rule.ruleClass
+            let name = target.rule.name
+
+            guard let ruleType = TopLevelRuleType(rawValue: kind) else {
+                if initializedConfig.baseConfig.targets.contains(name) {
+                    throw BazelTargetStoreError.unsupportedTargetType(target: name, type: kind)
+                }
+                // Ignore any targets that are not in our base config and aren't top-level: these are dependencies of our top-level targets
+                continue
+            }
+
+            topLevelTargetLabels.append(name)
```

Comment:
I wonder if we shouldn't do the `if initializedConfig.baseConfig.targets.contains(name)` check in the positive case as well to avoid unexpected `app -> app` dependencies from being picked up here. A couple of lines below we are already explicitly trying to avoid this when building the dependency graph, but it might be good to handle it here as well to match the old behavior